platforms/software/dependency-management/src/main/java/org/gradle/internal/resolve/resolver/FilteringResolvedArtifactSet.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.resolve.resolver;
+
+import org.gradle.api.Action;
+import org.gradle.api.attributes.AttributeContainer;
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ArtifactVisitor;
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ResolvableArtifact;
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ResolvedArtifactSet;
+import org.gradle.api.internal.artifacts.transform.TransformStepNode;
+import org.gradle.api.internal.file.FileCollectionInternal;
+import org.gradle.api.internal.file.FileCollectionStructureVisitor;
+import org.gradle.api.internal.tasks.TaskDependencyResolveContext;
+import org.gradle.internal.DisplayName;
+import org.gradle.internal.component.external.model.ImmutableCapabilities;
+import org.gradle.internal.operations.BuildOperationQueue;
+import org.gradle.internal.operations.RunnableBuildOperation;
+
+import java.util.function.Predicate;
+
+/**
+ * A {@link ResolvedArtifactSet} that applies a filter to the artifacts of a delegate {@link ResolvedArtifactSet}.
+ * <p>
+ * The filter is applied <strong>after</strong> build dependencies are calculated, meaning the filter is not
+ * applied to build dependencies. This is because the filter may be a function of the resolved artifact files,
+ * which are not known until after build dependencies are executed.
+ */
+class FilteringResolvedArtifactSet implements ResolvedArtifactSet {
+
+    private final ResolvedArtifactSet artifacts;
+    private final Predicate<ResolvableArtifact> filter;
+
+    public FilteringResolvedArtifactSet(ResolvedArtifactSet artifacts, Predicate<ResolvableArtifact> filter) {
+        this.artifacts = artifacts;
+        this.filter = filter;
+    }
+
+    @Override
+    public void visit(Visitor visitor) {
+        artifacts.visit(new Visitor() {
+            @Override
+            public FileCollectionStructureVisitor.VisitType prepareForVisit(FileCollectionInternal.Source source) {
+                return visitor.prepareForVisit(source);
+            }
+
+            @Override
+            public void visitArtifacts(Artifacts artifacts) {
+                visitor.visitArtifacts(new Artifacts() {
+                    @Override
+                    public void prepareForVisitingIfNotAlready() {
```

Comment:
```suggestion
                    public void prepareForVisitingIfNotReady() {
```

Reply:
I would like to avoid changes to `ResolvedArtifactSet` at the moment. This class is scary